### PR TITLE
fix: track version in fingerprint dep-info files

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Encodes the version information into Cargo's fingerprint dep-info files,
so that when the format encoding changes in the future,
Cargo understands a dep-info file was outdated and doesn't bother parsing it.

Since there was no version info encoded in the old format (call it v0),
to be compatible with older cargoes,
this PR works around it with a horrible hack.
It is explained in the doc comment of `EncodedDepInfo`.

### How should we test and review this PR?

This PR also adds a static file which is v0 depinfo,
to ensure newer cargoes give up parsing it successfully without allocation errors.
See the allocation failures in this CI job result: https://github.com/weihanglo/cargo/actions/runs/11601065840/job/32302943597

It is no harm to keep the v0 test file also fine with removing it from git.

### Additional information

Fixes rust-lang/cargo#14712

### CODE REVIEW

File: src/cargo/core/compiler/fingerprint/dep_info.rs [hunk 1]
```diff
@@ -84,7 +84,7 @@ pub enum DepInfoPathType {
 ///   | len of key | key bytes | value exists? | len of value | value bytes |
```
Comment (epage) [inline src/cargo/core/compiler/fingerprint/dep_info.rs:100]: Should we also have a sanity-check value on "len of key"  We likely should never have a dep info key that is larger than 10k for instance

Comment (epage) [reply src/cargo/core/compiler/fingerprint/dep_info.rs:100]: (and ture for any length we deal with)

File: src/cargo/core/compiler/fingerprint/dep_info.rs [hunk 2]
```diff
@@ -61,13 +64,26 @@ pub enum DepInfoPathType {
...
+///   Since [`EncodedDepInfo`] is just an optimization, to avoid adding any complexity,
+///   Cargo recognizes only one version of [`CURRENT_ENCODED_DEP_INFO_VERSION`].
+///   To support v0 (prior to checksum support in [`f4ca7390`]),
+///   which doesn't have this field, we do a hacky workaround. The current layout looks like
+///   ```text
+///   +----------------------------+---------+
+///   | [0x01 0x00 0x00 0x00 0xff] | version |
+///   +----------------------------+---------+
+///   ```
+///   The first 5 bytes will be interpreted as "one file trackced and invalid
```
Comment (epage) [inline src/cargo/core/compiler/fingerprint/dep_info.rs:84]: `trackced` -> `tracked`

File: src/cargo/core/compiler/fingerprint/dep_info.rs [hunk 3]
```diff
@@ -129,6 +152,16 @@ impl EncodedDepInfo {
         }
         return Some(EncodedDepInfo { files, env });
 
+        fn read_version(bytes: &mut &[u8]) -> Option<u8> {
+            let _size = read_usize(bytes)?;
+            let path_type = read_u8(bytes)?;
+            if path_type != u8::MAX {
+                // Old depinfo. Give up parsing it.
+                return None;
+            }
```
Comment (epage) [inline src/cargo/core/compiler/fingerprint/dep_info.rs:161]: imo this function deserves comments, even if its pointing back to `EncodedDepInfo`s comment

File: src/cargo/core/compiler/fingerprint/dep_info.rs (hunk 1 above)
Comment (weihanglo) [reply src/cargo/core/compiler/fingerprint/dep_info.rs:100]: Not sure. What if they do track that many environment variables?

An allocation error seems reasonable in that case. Cargo may crash in other places even it doesn't here.

File: src/cargo/core/compiler/fingerprint/dep_info.rs (hunk 3 above)
Comment (epage) [inline src/cargo/core/compiler/fingerprint/dep_info.rs:161]: I wonder if this should be split out into a `read_header` or `read_magic_marker` or something.  I found it confusing when I saw the doc comment and then immediately saw `read_version`.  I was then going to comment in `read_version` about how it was behaving

File: src/cargo/core/compiler/fingerprint/dep_info.rs [hunk 4]
```diff
@@ -667,4 +667,30 @@ mod encoded_dep_info {
...
+    // │000000d0│ 41 44 44 5f 42 49 4e 00 ┊ 08 00 00 00 54 46 5f 42 │ADD_BIN⋄┊•⋄⋄⋄TF_B│
+    // │000000e0│ 55 49 4c 44 00          ┊                         │UILD⋄   ┊        │
+    // └────────┴─────────────────────────┴─────────────────────────┴────────┴────────┘
+    #[test]
+    #[cfg_attr(not(target_os = "macos"), ignore = "OOM")]
+    fn parse_v0_fingerprint_dep_info() {
+        let data = include_bytes!("dep-lib-cargo_test_macro");
+        assert!(EncodedDepInfo::parse(&data[..]).is_none());
+    }
 }
```
Comment (epage) [inline src/cargo/core/compiler/fingerprint/dep_info.rs]: If this test isn't updated in the following commit, what is it we are expecting out of this test?

File: src/cargo/core/compiler/fingerprint/dep_info.rs [hunk 5]
```diff
@@ -129,6 +152,16 @@ impl EncodedDepInfo {
         }
         return Some(EncodedDepInfo { files, env });
 
+        fn read_version(bytes: &mut &[u8]) -> Option<u8> {
+            let _size = read_usize(bytes)?;
+            let path_type = read_u8(bytes)?;
+            if path_type != u8::MAX {
+                // Old depinfo. Give up parsing it.
+                return None;
```
Comment (epage) [inline src/cargo/core/compiler/fingerprint/dep_info.rs:160]: Should we have a test for this case?

File: src/cargo/core/compiler/fingerprint/dep_info.rs [hunk 6]
```diff
@@ -129,6 +155,18 @@ impl EncodedDepInfo {
         }
         return Some(EncodedDepInfo { files, env });
 
+        /// See [`EncodedDepInfo`] for why a magic marker exists.
+        fn read_magic_marker(bytes: &mut &[u8]) -> Option<()> {
+            let _size = read_usize(bytes)?;
+            let path_type = read_u8(bytes)?;
+            if path_type != u8::MAX {
+                // Old depinfo. Give up parsing it.
+                return None;
```
Comment (epage) [inline src/cargo/core/compiler/fingerprint/dep_info.rs:164]: nit: `return None;` could now be `None`

**Comment by epage** [comment]
Feel free to `r=` me now or after resolving the nit

File: src/cargo/core/compiler/fingerprint/dep_info.rs (hunk 1 above)
Comment (weihanglo) [reply src/cargo/core/compiler/fingerprint/dep_info.rs:103]: An alternative is <https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.try_with_capacity>.
(not yet stable though)

**Comment by weihanglo** [comment]
@bors r=epage

While none of the test covers the OOM cases described in <https://github.com/rust-lang/cargo/issues/14712>, <https://github.com/weihanglo/cargo/pull/69> shows both failed and successful builds before and after this patch.

**Pull request merged**